Update F46 to cover tables with structural elements that are marked presentational

mraccess77 opened the pull request:
Resolves #3565

mraccess77 committed:
Proposed update to F46 for presentational markup

mraccess77 committed:
Revert "Proposed update to F46 for presentational markup"

This reverts commit 3e81da13fe854b8812dd830a4cde0a59a4a9b5e6.

mraccess77 committed:
Proposed updated for presentational markup

detlevhfischer commented:
Looks good to me.

mbgower commented on techniques/failures/F46.html:
```diff
@@ -1,12 +1,12 @@
    </section><section id="description"><h2>Description</h2>
       <p>The objective of this technique is to describe a failure that occurs when a
-                        table used only for layout includes either th elements, a
+                        table used only for layout and is not marked presentational is used only for layout includes either th elements, a
```
```suggestion
                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, a
```

mbgower committed:
Update techniques/failures/F46.html

rewording to make the two considerations clearer

mbgower commented on techniques/failures/F46.html:
```diff
@@ -1,12 +1,12 @@
-                        table used only for layout includes either th elements, a
+                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, a
                             summary attribute, or a caption element. This
                         is a failure because it uses structural (or semantic) markup only for
```
```suggestion
                        is a failure because it exposes structural (or semantic) markup only for
```

mbgower committed:
Update techniques/failures/F46.html

rewording since the issue is no longer that it uses these elements but that it exposes them, by not overriding with presentation designation

mbgower commented on techniques/failures/F46.html:
```diff
@@ -1,14 +1,14 @@
+                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, a
                             summary attribute, or a caption element. This
-                        is a failure because it uses structural (or semantic) markup only for
+                        is a failure because it exposes structural (or semantic) markup only for
```
```suggestion
                        is a failure because it exposes structural (or semantic) markup in content only intended for
```

mbgower committed:
Update techniques/failures/F46.html

slight rewording

mbgower commented on techniques/failures/F46.html:
```diff
@@ -1,14 +1,14 @@
+                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, a
                             summary attribute, or a caption element. This
-                        is a failure because it uses structural (or semantic) markup only for
+                        is a failure because it exposes structural (or semantic) markup in content only intended for
```
```suggestion
                        is a failure because it exposes structural (or semantic) table markup in content only intended for
```

mbgower committed:
Update techniques/failures/F46.html

sneaking in the word "table" to clarify the kind of markup

mbgower commented on techniques/failures/F46.html:
```diff
@@ -1,14 +1,14 @@
-                        is a failure because it uses structural (or semantic) markup only for
+                        is a failure because it exposes structural (or semantic) table markup in content only intended for
                         presentation. The intent of the HTML and XHTML table elements is to present data. </p>
       <p>Although not commonly used in a layout table, the following structural markup would also be failures of Success Criterion 1.3.1 if used in a layout table: </p>
```
```suggestion
      <p>Although not commonly used in a layout table, the following structural markup would also be failures of Success Criterion 1.3.1 if exposed in a layout table: </p>
```

mbgower committed:
Update techniques/failures/F46.html

mbgower commented on techniques/failures/F46.html:
```diff
@@ -84,8 +85,9 @@
          <ul>
-            <li>If any check above is false, then this failure condition applies
-                                    and the content fails the Success Criterion.</li>
+            <LI>If check one above is true, then this failure condition does not apply</LI>
```
```suggestion
            <LI>If check #1 above is true, then this failure condition does not apply</LI>
```

mbgower commented on techniques/failures/F46.html:
```diff
@@ -84,8 +85,9 @@
-            <li>If any check above is false, then this failure condition applies
-                                    and the content fails the Success Criterion.</li>
+            <LI>If check one above is true, then this failure condition does not apply</LI>
+            <li>If check one above is false, and any subsequent check above is false, then this failure condition applies
```
```suggestion
            <li>If #1 is false, and any subsequent step is false, then this failure condition applies
```

mbgower commented on techniques/failures/F46.html:
```diff
@@ -72,6 +72,7 @@
                <p>If the table is used only to visually lay out elements within the
                                     content </p>
                <ol>
+                  <li>Check that the table is marked presentational by methods such as role none or presentation on the table element</li>
```
```suggestion
                  <li>Check that the table is marked presentational by methods such as role none or presentation on the table element.</li>
```

mbgower commented on techniques/failures/F46.html:
```diff
@@ -72,6 +72,7 @@
                <p>If the table is used only to visually lay out elements within the
                                     content </p>
```
```suggestion
                                    content:</p>
```

mbgower committed:
Update techniques/failures/F46.html

using what seems like the normal punctuation for lists in a test procedure

mbgower committed:
Update techniques/failures/F46.html

Adding terminal punctuation to each list element, which appears to be the style standard in the test procedures.

mbgower committed:
Update techniques/failures/F46.html

mbgower committed:
Update techniques/failures/F46.html

mbgower committed:
Update techniques/failures/F46.html

mgifford commented on techniques/failures/F46.html:
```diff
@@ -70,8 +70,9 @@
-                                    content </p>
+                                    content:</p>
                <ol>
+                  <li>Check that the table is marked presentational by methods such as role none or presentation on the table element.</li>
```
I'm just wondering if we should be clear that we're talking about role="none" and role="presentation".

It seems that the non-technical approach is both more confusing and more lengthy. Is there a reason not to be specific?

ljoakley replied:
I think that changing the text to reflect how the actual tag appears is the correct thing to do here.

mgifford commented on techniques/failures/F46.html:
```diff
@@ -84,8 +85,9 @@
          <ul>
-            <li>If any check above is false, then this failure condition applies
-                                    and the content fails the Success Criterion.</li>
+            <LI>If #1 is true, then this failure condition does not apply.</LI>
```
I like lower case \<li> elements, but that's pretty trivial.  
  
I think there must be an easiier way to phrase this. Whether #1 is true or false, this fails the SC. I think this statement is what it is trying to say.

mbgower replied:
I find the resulting 'double negative' a bit harder to parse in your suggestion.

The phrasing used in the test is the normal way such things are expressed in the failure techniques (although there is variation in the 100+ [failure techniques](https://www.w3.org/WAI/WCAG22/Techniques/#failures)).

mgifford replied:
There's a lot of content in WCAG to maintain...  And ya, double negatives are annoying.   
  
Could it be simplified to:   
  "If #1 is true, then this does not apply."

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -70,8 +70,9 @@
-                                    content </p>
+                                    content:</p>
                <ol>
+                  <li>Check that the table is marked presentational by methods such as role none or presentation on the table element.</li>
```
redoing this @mbgower as you can't use markdown here

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -70,8 +70,9 @@
-                                    content </p>
+                                    content:</p>
                <ol>
+                  <li>Check that the table is marked presentational by methods such as role none or presentation on the table element.</li>
```
```suggestion
                  <li>Check that the table is marked presentational by methods such as <code>role="none"</code> or <code>role="presentation"</code> on the table element.</li>
```

patrickhlauke committed:
Update techniques/failures/F46.html

patrickhlauke commented:
@mbgower made the discussed change of adding `<code>` around `role="none"` and `role="presentation"` (using markup rather than markdown)

scottaohara reviewed (changes requested):
looking at this from the wcag 2.x taskforce meeting, i raised a concern that while the procedure that was added to determine if a table is a layout table or not - there are many other factors that could result in a table's exposed role being changed, or not.

it might be a whole lot simpler to instead write a procedure that instructs people to check the exposed role for the table using browser dev tools - and then mention that there are many factors (including browser heuristics) that could result in a table being exposed as such, or not).

awkawk commented on techniques/failures/F46.html:
```diff
@@ -70,8 +70,9 @@
                                     table element</li>
             <li>
                <p>If the table is used only to visually lay out elements within the
```
We've done the setup line in a different way usually - e.g., "For each table used only to visually lay out elements within the page:"

awkawk commented on techniques/failures/F46.html:
```diff
@@ -84,8 +85,9 @@
-            <li>If any check above is false, then this failure condition applies
-                                    and the content fails the Success Criterion.</li>
+            <LI>If #1 is true, then this failure condition does not apply.</LI>
+            <li>If #1 is false, and any subsequent step is false, then this failure condition applies
```
If there are headings in a table that is marked presentational, what do the browsers expose to AT about that? I assume they don't convey table semantics in that case.

mraccess77 replied:
I would assume that was well - but I'm not clear what you are saying needs to be updated.

awkawk replied:
I guess I'm finding the construction confusing. Perhaps we just have one expected result which would be:

If 1 and either 2 or 3 are false, then the failure condition applies.

ljoakley replied:
> If there are headings in a table that is marked presentational, what do the browsers expose to AT about that? I assume they don't convey table semantics in that case.

Yes, a screen reader will skip over a header labeled role-presentational. However, if a table has headers, that makes it a data table and now the rules for data tables apply.

mbgower commented:
@mraccess77 are you able to take this feedback and produce something? @scottaohara is happy to assist.

mraccess77 commented:
Do we have any example techniques were we have looked at the accessibility tree in the browser? This seems like a new approach from what our technique typically look like and I want to make sure we are consistent.  Can we assume that is the new way to go for all techniques?  For example, should all of our image with alternative text techniques be rewritten to cover the accessible name?

patrickhlauke commented:
i believe @scottaohara's idea - if i follow it correctly - would be to generalise the wording to something more like:

```
Failure of Success Criterion 1.3.1 due to using th elements, caption elements, or non-empty summary attributes in layout tables that are still exposed as data tables
```

and then adding an explanation that is more generic, such as "unless the semantics of the table have been removed - for instance, using `role="none"` or `role="presentation"` on the `<table>` element - they will still be exposed as data tables to screen readers" or similar...

scottaohara commented:
linking related issues (layout tables) https://github.com/w3c/wcag/issues/3524

ljoakley commented:
I find it interesting in reading through the definition of tables in both MDN and W3schools they both are using data tables and make no distinct between layout tables and data tables.

scottaohara commented:
@ljoakley specific things you can link to?

i ask because there are layout tables in the scope of accessibility, and then there are also layout-table CSS properties

bruce-usab commented:
Discussed 9/13, moved to To Do.

fstrr commented on techniques/failures/F46.html:
```diff
@@ -1,16 +1,16 @@
                     caption elements, or non-empty summary attributes in
-                    layout tables</h1><section class="meta"><p class="id">ID: F46</p><p class="technology">Technology: failures</p><p class="type">Type: Failure</p></section><section id="applicability"><h2>When to Use</h2>
+                    layout tables that are not marked presentational</h1><section class="meta"><p class="id">ID: F46</p><p class="technology">Technology: failures</p><p class="type">Type: Failure</p></section><section id="applicability"><h2>When to Use</h2>
       <p>HTML and XHTML</p>
```
```suggestion
      <p>HTML</p>
```

fstrr commented on techniques/failures/F46.html:
```diff
@@ -1,16 +1,16 @@
                             summary attribute, or a caption element. This
-                        is a failure because it uses structural (or semantic) markup only for
+                        is a failure because it exposes structural (or semantic) table markup in content only intended for
                         presentation. The intent of the HTML and XHTML table elements is to present data. </p>
```
```suggestion
                        presentation. The intent of the HTML table element is to present tabular data.</p>
```

fstrr committed:
Update techniques/failures/F46.html

fstrr committed:
Update techniques/failures/F46.html

fstrr commented on techniques/failures/F46.html:
```diff
@@ -1,16 +1,16 @@
    </section><section id="description"><h2>Description</h2>
       <p>The objective of this technique is to describe a failure that occurs when a
-                        table used only for layout includes either th elements, a
+                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, a
                             summary attribute, or a caption element. This
```
```suggestion
                            , or a caption element. This
```

fstrr replied:
```suggestion
                        table that is used only for layout is both 1) not marked as presentational and 2) includes either th elements, or a caption element. This
```

fstrr commented on techniques/failures/F46.html:
```diff
@@ -1,16 +1,16 @@
 <!DOCTYPE html><html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml"><head><title>Failure of Success Criterion 1.3.1 due to using th elements,
                     caption elements, or non-empty summary attributes in
```
```suggestion
<!DOCTYPE html><html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml"><head><title>Failure of Success Criterion 1.3.1 due to using th elements or
                    caption elements in
```

fstrr commented on techniques/failures/F46.html:
```diff
@@ -1,16 +1,16 @@
                     caption elements, or non-empty summary attributes in
-                    layout tables</title><link rel="stylesheet" type="text/css" href="../../css/editors.css" class="remove"></link></head><body><h1>Failure of Success Criterion 1.3.1 due to using th elements,
+                    layout tables that are not marked presentational</title><link rel="stylesheet" type="text/css" href="../../css/editors.css" class="remove"></link></head><body><h1>Failure of Success Criterion 1.3.1 due to using th elements,
                     caption elements, or non-empty summary attributes in
```
```suggestion
                    caption elements in
```

fstrr commented:
[The summary attribute for tables is obsolete](https://html.spec.whatwg.org/multipage/obsolete.html), so would suggest updating the name of the technique to remove references to it (we've removed other `summary`-related content before. This includes re-naming the Technique. I've added suggestions to the PR.

I've also removed references to XHTML.

patrickhlauke committed:
Update techniques/failures/F46.html

Co-authored-by: Detlev Fischer <df@3needs.org>

patrickhlauke committed:
Update techniques/failures/F46.html

Co-authored-by: Detlev Fischer <df@3needs.org>

patrickhlauke committed:
Update techniques/failures/F46.html

Co-authored-by: Francis Storr <francis.storr@intel.com>

patrickhlauke committed:
Update techniques/failures/F46.html

Co-authored-by: Francis Storr <francis.storr@intel.com>

patrickhlauke committed:
Update techniques/failures/F46.html

Co-authored-by: Francis Storr <francis.storr@intel.com>

patrickhlauke committed:
Merge branch 'main' into main

patrickhlauke committed:
Merge branch 'main' into main

patrickhlauke committed:
Merge branch 'main' into main

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -1,22 +1,30 @@
+         <li><code>scope</code> attributes</li>
       </ul>
       <p>Assistive technologies use the structure of an HTML table to present data to
                         the user in a logical manner. The th element is used to mark
```
```suggestion
                        the user in a logical manner. The <code>th</code> element is used to mark
```

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -28,8 +36,7 @@
                             summary attribute information available to users. The
```
```suggestion
                            <code>summary</code> attribute information available to users. The
```

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -28,8 +36,7 @@
-            Although WCAG 2 does not prohibit the use of layout tables, CSS-based layouts are recommended in order to retain the defined semantic meaning of the HTML table elements and to conform to the coding practice of separating presentation from content. 
+      <p>Although WCAG 2 does not prohibit the use of layout tables, CSS-based layouts are recommended in order to retain the defined semantic meaning of the HTML table elements and to conform to the coding practice of separating presentation from content. 
             When a table is used
                         for layout purposes the th element should not be used. Since
```
```suggestion
                        for layout purposes the <code>th</code> element should not be used. Since
```

patrickhlauke commented on techniques/failures/F46.html:
```suggestion
                        a <code>summary</code> attribute and do not use the <code>summary</code>
```

patrickhlauke commented on techniques/failures/F46.html:
```suggestion
                        navigating the content via a screen reader. Empty <code>summary</code>
```

patrickhlauke commented on techniques/failures/F46.html:
```suggestion
                                title as <code>th</code>.</p>
```

patrickhlauke committed:
Apply suggestions from code review

Co-authored-by: Patrick H. Lauke <redux@splintered.co.uk>

patrickhlauke commented on techniques/failures/F46.html:
```diff
@@ -1,44 +1,51 @@
+                        the user in a logical manner. The <code>th</code> element is used to mark
                         the column and row headers of the table. A screen reader uses the
                         information in th elements to speak the header information that
                         changes as the user navigates the table. The summary attribute on the
```
```suggestion
                        changes as the user navigates the table. The <code>summary</code> attribute on the
```

patrickhlauke committed:
Apply suggestion from @patrickhlauke

patrickhlauke committed:
Remove mention of deprecates `summary` attribute

patrickhlauke commented:
I started fixing this PR up a bit, but now taking a step back I'm wondering: what is this failure actually trying to achieve? Isn't it already a failure to have a layout table that is still exposed as a data table, in general? i.e. would we not already fail a table that is used for layout but does NOT include something like `role="presentation"` or `role="none"` or similar?

IF a table is exposed as presentational, then it shouldn't make a difference if there are `th` or similar structural elements, as they won't be exposed as such.

so really the failure currently seems to be a weird mitigation - IF you're using a layout table, but you didn't mark it as presentational (you already failed?), make sure you don't use structural elements (which would fail...harder?)

/cc @scottaohara

patrickhlauke committed:
First pass cleaning up F46 further

bruce-usab commented:
> Isn't it already a failure to have a layout table that is still exposed as a data table, in general?

Yes.

> i.e. would we not already fail a table that is used for layout but does NOT include something like role="presentation" or role="none" or similar?

No.  Historically, a layout table merely means avoiding `TH` or similar.  See WCAG 1.0 (!) [Create tables that transform gracefully](https://www.w3.org/WAI/GL/WCAG10/wcag10.html#gl-table-markup).

patrickhlauke commented:
@bruce-usab your "Yes" and then "No" are...in contradiction?

a layout table is a table used for layout. every table is exposed as a table by default (though browsers often use heuristics to suppress this themselves, but not sure they're documented officially anywhere?). you only suppress the semantics of data table when you explicitly mark it as such. in wcag 1.0 days the ability to do this probably didn't exist (with `role="presentation"` or similar)

bruce-usab commented:
> @bruce-usab your "Yes" and then "No" are...in contradiction?

If a table is being used for layout, but using `TH` to emphasize certain cells (and not as a label for row/column/row group) then that fails 1.3.1.  So **yes**, in general, it’s already a failure to have a layout table that is still exposed as a data table.

> you only suppress the semantics of data table when you explicitly mark it as such. in wcag 1.0 days the ability to do this probably didn't exist (with `role="presentation"` or similar)

Each of these data points supports my conclusion that, **no**, one doesn’t need anything as explicit as role="presentation" for layout tables.  It could be written up as a best practice.  And maybe use of role="presentation" _excuses_ using `TH` for emphasis (rather than as a label) and creates a loophole?

I don't think 1.3.1 is generally relevant to layout tables, except for an author misusing code (e.g., `TH` for emphasis, not as a label).  The main concern for layout tables is SC 1.3.2 Meaningful Sequence.

One pernicious pattern I’ve come across (more than once) is where the headers are one table and the data is in a second table immediately following.  Visually it looks like a single data table.  [The pages were being generated from some legacy databases.  (Maybe Oracle, but it could have been COBOL for all I know.)]  The assertion, by the owner, that these were layout tables was obviously disingenuous, but it’s the sort of thing which was easier to fail under WCAG 1.0 (and original 508).  I couldn’t point to 3.3.2 Labels or Instructions because the context wasn’t input, only an output report.  1.3.1 is so overloaded, but I don’t have any great ideas for improving that.

patrickhlauke commented:
> one doesn’t need anything as explicit as role="presentation" for layout tables

every table is by its very use/nature a data table, unless explicitly marked as presentational, no? i.e. screen readers will announce the presence of a table with X rows and Y columns, for instance.

or is your definition for layout tables more lax?

and again, if the group has the same lax view ("it's ok to use a table structure for layout purposes, without explicitly marking it/conveying it as presentational/neutralising table semantics, as long as no other semantic structures like `th` are used") then that needs to be explicitly documented in a note/exception

patrickhlauke committed:
Merge branch 'main' into main
